### PR TITLE
feat(chore): germ stage implementation of config overrides

### PR DESCRIPTION
**Pull request by kevinslin**
feat(chore): germ stage implementation of config overrides

This is an early build of introducing [[46 Config Overrides|dendron://dendron.docs/rfc.46-config-overrides]]

It allows a custom `dendronrc.yml` file, located either inside the `$WS_ROOT` or the `$HOME` folder to be merged with the values in `dendron.yml`.

Most immediately, the usecase is to support dendron team members to easily add their own local vaults to the team workspace.

Docs: https://github.com/dendronhq/dendron-docs/pull/27

### Other Changes
- the `Extensions.test` file is getting quite large and conflates activating the extension and testing `SetupWorkspace`
    - split `Extensions.test` into 4 tests
        - Extension.test: test regular activation logic
        - Extension-PostInstall: test plugin install logic
        - Extension-PostUpgrade: test version upgrade logic
    - add `SetupWorkspace.test`
    - refactor some modules into shared test utilities

### Metrics
- adds the following metrics (will be synced with airtable once approved)

- EventProperties (part of Initialize Workspace):
    - hasLocalConfig, boolean, indicates if a local config is present
    - numLocalConfigVaults, number, number of vaults that are loaded from local config

### Post Approval tasks
- [ ] async about extensions.test
- [ ] update metrics

***

# Dendron Extended PR Checklist

### Basics

- [x] code should follow [Code Conventions](https://wiki.dendron.so/notes/773e0b5a-510f-4c21-acf4-2d1ab3ed741e.html)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](https://wiki.dendron.so/notes/pMS27sHxbWeKMoPRrWEzs.html).
- [x] sticking to existing conventions instead of creating new ones 
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended

- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed 
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics

- [x] if you are adding analytics related changes, make sure you [Document](https://wiki.dendron.so/notes/8ThPaB9iXXm2Szk3C9kFt.html) changes in airtable

### Extended

- [x] can we track the performance of this change to know if it is _successful_? 
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

## 

### Basics

- [x] [Write Tests](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html) 
- [x] [Confirm existing tests pass](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html)
- [x] [Confirm manual testing](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](https://wiki.dendron.so/notes/99q7A73uGmCwu2KvSHZro.html)

### Extended

NA

### Basics

- [x] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

## 

### Basics

- [x] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](https://wiki.dendron.so/notes/3489b652-cd0e-4ac8-a734-08094dc043eb.html) or [Packages](https://wiki.dendron.so/notes/32cdd4aa-d9f6-4582-8d0c-07f64a00299b.html)

## 

### Extended

- [x]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](https://wiki.dendron.so/notes/iZ8vpfY0n1E3Nq3IC1Y7X.html)
  - eg. [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

File: packages/engine-server/src/workspace/service.ts [hunk 1]
```diff
@@ -874,9 +879,8 @@ export class WorkspaceService implements Disposable, IWorkspaceService {
   }
 
   createConfig() {
-    // This line actually does something: it will create a config if one doesn't exist.
     // eslint-disable-next-line no-unused-expressions
```
Comment (jonathanyeung) [inline packages/engine-server/src/workspace/service.ts:882]: nit: in this case, we can also remove the eslint line right?

**Review by jonathanyeung — APPROVED** [review]
Looks good for a dogfood test.  Just minor comments / q's.

File: packages/engine-server/src/config.ts [hunk 2]
```diff
@@ -174,6 +188,73 @@ export class DConfig {
...
+    wsRoot: string
+  ): RespV3<IntermediateDendronConfig> {
+    const wsPath = path.join(wsRoot, CONSTANTS.DENDRON_LOCAL_CONFIG_FILE);
+    const globalPath = path.join(
+      os.homedir(),
+      CONSTANTS.DENDRON_LOCAL_CONFIG_FILE
+    );
+    let foundPath: string | undefined;
+
+    if (fs.existsSync(globalPath)) {
```
[the remark this answers is not in the document]
Comment (kevinslin) [reply packages/engine-server/src/config.ts:204]: yep, its one or the other. its in line with how rc files currently work in tools like `eslint`. can change in the future but opt for simple behavior for now

File: packages/engine-server/src/workspace/service.ts (hunk 1 above)
Comment (kevinslin) [reply packages/engine-server/src/workspace/service.ts:882]: you're right. fixed!

**Pull request merged**